components/gfx/font_context.rs

```diff
@@ -9,95 +9,64 @@ use std::sync::Arc;
 
 use app_units::Au;
 use fnv::FnvHasher;
-use log::debug;
+use ipc_channel::ipc::{self, IpcSender};
+use log::{debug, trace};
 use malloc_size_of::{MallocSizeOf, MallocSizeOfOps};
 use malloc_size_of_derive::MallocSizeOf;
-use parking_lot::{Mutex, RwLock};
+use net_traits::request::{Destination, Referrer, RequestBuilder};
+use net_traits::{fetch_async, CoreResourceThread, FetchResponseMsg, ResourceThreads};
+use parking_lot::{Mutex, ReentrantMutex, RwLock};
 use servo_arc::Arc as ServoArc;
 use style::computed_values::font_variant_caps::T as FontVariantCaps;
+use style::font_face::{FontFaceSourceFormat, FontFaceSourceFormatKeyword, Source, UrlSource};
+use style::media_queries::Device;
 use style::properties::style_structs::Font as FontStyleStruct;
-use webrender_api::{FontInstanceFlags, FontInstanceKey};
-
-use crate::font::{Font, FontDescriptor, FontFamilyDescriptor, FontGroup, FontRef};
-use crate::font_cache_thread::FontIdentifier;
-use crate::font_template::{FontTemplateRef, FontTemplateRefMethods};
+use style::shared_lock::SharedRwLockReadGuard;
+use style::stylesheets::{Stylesheet, StylesheetInDocument};
+use style::Atom;
+use url::Url;
+
+use crate::font::{
+    Font, FontDescriptor, FontFamilyDescriptor, FontFamilyName, FontGroup, FontRef, FontSearchScope,
+};
+use crate::font_cache_thread::{
+    CSSFontFaceDescriptors, FontIdentifier, FontSource, LowercaseString,
+};
+use crate::font_store::{CrossThreadFontStore, CrossThreadWebRenderFontStore};
+use crate::font_template::{FontTemplate, FontTemplateRef, FontTemplateRefMethods};
 #[cfg(target_os = "macos")]
 use crate::platform::core_text_font_cache::CoreTextFontCache;
 
 static SMALL_CAPS_SCALE_FACTOR: f32 = 0.8; // Matches FireFox (see gfxFont.h)
 
-pub trait FontSource {
-    fn get_font_instance(
-        &mut self,
-        font_identifier: FontIdentifier,
-        size: Au,
-        flags: FontInstanceFlags,
-    ) -> FontInstanceKey;
-    fn find_matching_font_templates(
-        &mut self,
-        descriptor_to_match: &FontDescriptor,
-        family_descriptor: FontFamilyDescriptor,
-    ) -> Vec<FontTemplateRef>;
-}
-
 /// The FontContext represents the per-thread/thread state necessary for
 /// working with fonts. It is the public API used by the layout and
 /// paint code. It talks directly to the font cache thread where
 /// required.
-#[derive(Debug)]
 pub struct FontContext<S: FontSource> {
-    font_source: Mutex<S>,
-
-    // TODO: The font context holds a strong ref to the cached fonts
-    // so they will never be released. Find out a good time to drop them.
-    // See bug https://github.com/servo/servo/issues/3300
-    font_cache: RwLock<HashMap<FontCacheKey, Option<FontRef>>>,
-    font_template_cache: RwLock<HashMap<FontTemplateCacheKey, Vec<FontTemplateRef>>>,
-    font_group_cache:
-        RwLock<HashMap<FontGroupCacheKey, Arc<RwLock<FontGroup>>, BuildHasherDefault<FnvHasher>>>,
+    font_source: Arc<ReentrantMutex<S>>,
```

Comment:
We should be able to remove the Arc around `font_source` and `resource_threads`.

I'm able to compile and run servo with this patch:
```
diff --git a/components/gfx/font_context.rs b/components/gfx/font_context.rs
index 560247efe7..0aff4abde2 100644
--- a/components/gfx/font_context.rs
+++ b/components/gfx/font_context.rs
@@ -44,8 +44,8 @@ static SMALL_CAPS_SCALE_FACTOR: f32 = 0.8; // Matches FireFox (see gfxFont.h)
 /// paint code. It talks directly to the font cache thread where
 /// required.
 pub struct FontContext<S: FontSource> {
-    font_source: Arc<ReentrantMutex<S>>,
-    resource_threads: Arc<ReentrantMutex<CoreResourceThread>>,
+    font_source: ReentrantMutex<S>,
+    resource_threads: ReentrantMutex<CoreResourceThread>,
     cache: CachingFontSource<S>,
     web_fonts: CrossThreadFontStore,
     webrender_font_store: CrossThreadWebRenderFontStore,
@@ -60,11 +60,10 @@ impl<S: FontSource> MallocSizeOf for FontContext<S> {
 impl<S: FontSource> FontContext<S> {
     pub fn new(font_source: S, resource_threads: ResourceThreads) -> FontContext<S> {
         #[allow(clippy::default_constructed_unit_structs)]
-        let font_source = Arc::new(ReentrantMutex::new(font_source));
         FontContext {
-            font_source: font_source.clone(),
-            resource_threads: Arc::new(ReentrantMutex::new(resource_threads.core_thread)),
-            cache: CachingFontSource::new(font_source),
+            font_source: ReentrantMutex::new(font_source.clone()),
+            resource_threads: ReentrantMutex::new(resource_threads.core_thread.clone()),
+            cache: CachingFontSource::new(font_source.clone()),
             web_fonts: Arc::new(RwLock::default()),
             webrender_font_store: Arc::new(RwLock::default()),
         }
@@ -508,7 +507,7 @@ impl<FCT: FontSource + Send + 'static> RemoteWebFontDownloader<FCT> {
 
 #[derive(Default)]
 pub struct CachingFontSource<FCT: FontSource> {
-    font_cache_thread: Arc<ReentrantMutex<FCT>>,
+    font_cache_thread: ReentrantMutex<FCT>,
     fonts: RwLock<HashMap<FontCacheKey, Option<FontRef>>>,
     templates: RwLock<HashMap<FontTemplateCacheKey, Vec<FontTemplateRef>>>,
     resolved_font_groups:
@@ -516,9 +515,9 @@ pub struct CachingFontSource<FCT: FontSource> {
 }
 
 impl<FCT: FontSource> CachingFontSource<FCT> {
-    fn new(font_cache_thread: Arc<ReentrantMutex<FCT>>) -> Self {
+    fn new(font_cache_thread: FCT) -> Self {
         Self {
-            font_cache_thread,
+            font_cache_thread: ReentrantMutex::new(font_cache_thread),
             fonts: Default::default(),
             templates: Default::default(),
             resolved_font_groups: Default::default(),
```

Reply:
Nice! I've gone ahead and integrated your changes.